CLI - dotnet workload restore fails on solution if there is a .esproj project

Similar to https://github.com/dotnet/sdk/issues/42394:

Running `dotnet workload restore` in a solution that has an `.esproj` project results in `error MSB4057: The target "_GetRequiredWorkloads" does not exist in the project.`,

# Environment
- happens on both Azure pipelines and local pc
- .NET SDK 9.0.100-rc.2.24474.11

# Azure Pipelines logs:

## select .NET SDK:
```
Starting: use dotnet sdk from global.json
==============================================================================
Task         : Use .NET Core
Description  : Acquires a specific version of the .NET Core SDK from the internet or the local cache and adds it to the PATH. Use this task to change the version of .NET Core used in subsequent tasks. Additionally provides proxy support.
Version      : 2.247.1
Author       : Microsoft Corporation
Help         : https://aka.ms/AA4xgy0
==============================================================================
Found version 9.0.100-rc.2.24474.11 in channel 9.0 for user specified version spec: 9.0.100-rc.2.24474.11
Getting URL to download .NET Core sdk version: 9.0.100-rc.2.24474.11.
Detecting OS platform to find correct download package for the OS.
get-os-distro: Error: Distribution specific OS name and version could not be detected: UName = Linux
/agent/_work/_tasks/UseDotNet_b0ce7256-7898-45d3-9cb5-176b752bfea6/2.247.1/externals/get-os-distro.sh
Primary:linux-x64
Legacy:-x64
Detected platform (Primary): linux-x64
Detected platform (Legacy): -x64
Version 9.0.100-rc.2.24474.11 was not found in cache.
Downloading: https://download.visualstudio.microsoft.com/download/pr/202e929a-e985-4eab-a78a-d7159fc204e4/0c85219d441cd3bbffd4fb65b7e36fe5/dotnet-sdk-9.0.100-rc.2.24474.11-linux-x64.tar.gz
Extracting downloaded package /agent/_work/_temp/7d050235-809f-4e8e-ab08-54fcbea01af6.
Extracting archive
/usr/bin/tar xC /agent/_work/_temp/17ca47b8-6a6d-4d73-8adf-9e0db0246ae7 -f /agent/_work/_temp/7d050235-809f-4e8e-ab08-54fcbea01af6

Successfully installed .NET Core sdk version 9.0.100-rc.2.24474.11.
Creating global tool path and pre-pending to PATH.
Finishing: use dotnet sdk from global.json
```

## dotnet restore workloads:

```
Starting: Restore workloads
==============================================================================
Task         : .NET Core
Description  : Build, test, package, or publish a dotnet application, or run a custom dotnet command
Version      : 2.247.3
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/build/dotnet-core-cli
==============================================================================
Info: .NET Core SDK/runtime 2.2 and 3.0 are now End of Life(EOL) and have been removed from all hosted agents. If you're using these SDK/runtimes on hosted agents, kindly upgrade to newer versions which are not EOL, or else use UseDotNet task to install the required version.
/opt/hostedtoolcache/dotnet/dotnet workload restore

Skipping NuGet package signature verification.
Skipping NuGet package signature verification.
Updated advertising manifest microsoft.net.sdk.macos.
Updated advertising manifest microsoft.net.sdk.android.
Updated advertising manifest microsoft.net.sdk.tvos.
Updated advertising manifest microsoft.net.sdk.maccatalyst.
Updated advertising manifest microsoft.net.sdk.maui.
Updated advertising manifest microsoft.net.workload.emscripten.net8.
Updated advertising manifest microsoft.net.workload.emscripten.current.
Updated advertising manifest microsoft.net.workload.emscripten.net6.
Updated advertising manifest microsoft.net.workload.mono.toolchain.current.
Updated advertising manifest microsoft.net.sdk.ios.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net8.
Updated advertising manifest microsoft.net.sdk.aspire.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net6.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net7.
Updated advertising manifest microsoft.net.workload.emscripten.net7.
Installing workload manifest microsoft.net.sdk.android version 35.0.0-rc.2.152...
Installing workload manifest microsoft.net.sdk.ios version 18.0.9600-net9-rc2...
Installing workload manifest microsoft.net.sdk.maccatalyst version 18.0.9600-net9-rc2...
Installing workload manifest microsoft.net.sdk.macos version 15.0.9600-net9-rc2...
Installing workload manifest microsoft.net.sdk.maui version 9.0.0-rc.2.24503.2...
Installing workload manifest microsoft.net.sdk.tvos version 18.0.9600-net9-rc2...
Installing workload manifest microsoft.net.sdk.aspire version 8.2.2...
No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.
Garbage collecting for SDK feature band(s) 9.0.100-rc.2...

Successfully updated workload(s): .

__________________________________________________
Project "/agent/_work/1/s/src/MyProject.Server/MyProject.Server.csproj" (_GetRequiredWorkloads target(s)):

__________________________________________________
Project "/agent/_work/1/s/src/MyProject.Server/MyProject.Server.csproj" is building "/agent/_work/1/s/src/MyProject.client/MyProject.client.esproj" (_GetRequiredWorkloads target(s)):

/agent/_work/1/s/src/MyProject.client/MyProject.client.esproj : error MSB4057: The target "_GetRequiredWorkloads" does not exist in the project.
Done building project "MyProject.client.esproj" -- FAILED.
Failed to restore workload for project /agent/_work/1/s/src/MyProject.Server/MyProject.Server.csproj: Failed to run MSBuild Target _GetRequiredWorkloads.

##[error]Error: The process '/opt/hostedtoolcache/dotnet/dotnet' failed with exit code 1
Info: Azure Pipelines hosted agents have been updated and now contain .Net 5.x SDK/Runtime along with the older .Net Core version which are currently lts. Unless you have locked down a SDK version for your project(s), 5.x SDK might be picked up which might have breaking behavior as compared to previous versions. You can learn more about the breaking changes here: https://docs.microsoft.com/en-us/dotnet/core/tools/ and https://docs.microsoft.com/en-us/dotnet/core/compatibility/ . To learn about more such changes and troubleshoot, refer here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/dotnet-core-cli?view=azure-devops#troubleshooting
##[error]Dotnet command failed with non-zero exit code on the following projects : [ '' ]
Finishing: Restore workloads
```

Workaround: add this to your project

```
  <Target Name="_GetRequiredWorkloads">
    <!-- This target is required to avoid the error MSB4057 -> myra-portal.esproj : error MSB4057: The target "_GetRequiredWorkloads" does not exist in the project. -->
  </Target>
```

Workaround: add this to your project

```
  <Target Name="_GetRequiredWorkloads">
    <!-- This target is required to avoid the error MSB4057 -> MyProject.client.esproj : error MSB4057: The target "_GetRequiredWorkloads" does not exist in the project. -->
  </Target>
```

--- a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.ImportWorkloads.targets
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.ImportWorkloads.targets
@@ -42,14 +42,17 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
   <!--  This target is not part of the build. Only used by dotnet workload restore command. Global property "SkipResolvePackageAssets"
-        need to be set to "true" to avoid requiring restore (which would likely fail if the required workloads aren't already installed).-->
+        need to be set to "true" to avoid requiring restore (which would likely fail if the required workloads aren't already installed).
+        In addition, since this is a target that is called on potentially-unsupported project types like esproj, we need to not fail
+        if the Target is missing. -->
   <Target Name="_GetRequiredWorkloads" DependsOnTargets="GetSuggestedWorkloads;PrepareProjectReferences" Returns="@(_ResolvedSuggestedWorkload)">
     <MSBuild
       Projects="@(_MSBuildProjectReferenceExistent)"
       Targets="_GetRequiredWorkloads"
       BuildInParallel="$(BuildInParallel)"
       Condition="'%(_MSBuildProjectReferenceExistent.BuildReference)' == 'true' and '@(_MSBuildProjectReferenceExistent)' != ''"
-      RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove)$(_GlobalPropertiesToRemoveFromProjectReferences);TargetFramework;TargetFrameworks">
+      RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove)$(_GlobalPropertiesToRemoveFromProjectReferences);TargetFramework;TargetFrameworks"
+      SkipNonexistentProjects="true" >
       <Output TaskParameter="TargetOutputs" ItemName="SuggestedWorkloadFromReference"/>
     </MSBuild>
 

--- a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.ImportWorkloads.targets
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.ImportWorkloads.targets
@@ -52,7 +52,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       BuildInParallel="$(BuildInParallel)"
       Condition="'%(_MSBuildProjectReferenceExistent.BuildReference)' == 'true' and '@(_MSBuildProjectReferenceExistent)' != ''"
       RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove)$(_GlobalPropertiesToRemoveFromProjectReferences);TargetFramework;TargetFrameworks"
-      SkipNonexistentProjects="true" >
+      SkipNonexistentTargets="true" >
       <Output TaskParameter="TargetOutputs" ItemName="SuggestedWorkloadFromReference"/>
     </MSBuild>
 

--- a/test/TestAssets/TestProjects/ProjectWithEsProjReference/App.csproj
+++ b/test/TestAssets/TestProjects/ProjectWithEsProjReference/App.csproj
@@ -1,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="./Lib.esproj" />
+  </ItemGroup>
+</Project>

--- a/test/TestAssets/TestProjects/ProjectWithEsProjReference/Lib.esproj
+++ b/test/TestAssets/TestProjects/ProjectWithEsProjReference/Lib.esproj
@@ -1,0 +1,1 @@
+<Project Sdk="Microsoft.VisualStudio.JavaScript.SDK/1.0.1948920" />

--- a/test/TestAssets/TestProjects/ProjectWithEsProjReference/package.json
+++ b/test/TestAssets/TestProjects/ProjectWithEsProjReference/package.json
@@ -1,0 +1,11 @@
+{
+  "name": "lib",
+  "version": "1.0.0",
+  "description": "A test project to verify .NET SDK compatibility for workload restore and non-Managed projects",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "MIT"
+}

--- a/test/dotnet-workload-restore.Tests/GivenDotnetWorkloadRestore.cs
+++ b/test/dotnet-workload-restore.Tests/GivenDotnetWorkloadRestore.cs
@@ -10,6 +10,7 @@ public class GivenDotnetWorkloadRestore : SdkTest
     }
 
     public static string DcProjAssetName = "SolutionWithAppAndDcProj";
+    public static string TransitiveReferenceNoWorkloadsAssetName = "ProjectWithEsProjReference";
 
     [Fact]
     public void ProjectsThatDoNotSupportWorkloadsAreNotInspected()
@@ -25,6 +26,23 @@ public class GivenDotnetWorkloadRestore : SdkTest
         .Execute()
         .Should()
         // if we did try to restore the dcproj in this TestAsset we would fail, so passing means we didn't!
+        .Pass();
+    }
+
+    [Fact]
+    public void ProjectsThatDoNotSupportWorkloadsAndAreTransitivelyReferencedDoNotBreakTheBuild()
+    {
+        var projectPath =
+            _testAssetsManager
+                .CopyTestAsset(TransitiveReferenceNoWorkloadsAssetName)
+                .WithSource()
+                .Path;
+
+        new DotnetWorkloadCommand(Log, "restore")
+        .WithWorkingDirectory(projectPath)
+        .Execute()
+        .Should()
+        // if we did try to restore the esproj in this TestAsset we would fail, so passing means we didn't!
         .Pass();
     }
 }

